test/hummingbot/connector/derivative/phemex_perpetual/test_phemex_perpetual_api_order_book_data_source.py

```diff
@@ -0,0 +1,583 @@
+        regex_url = re.compile(f"^{url}".replace(".", r"\.").replace("?", r"\?"))
+
+        mock_api.get(regex_url, exception=asyncio.CancelledError)
+
+        msg_queue: asyncio.Queue = asyncio.Queue()
+
+        with self.assertRaises(asyncio.CancelledError):
+            self.listening_task = self.ev_loop.create_task(
+                self.data_source.listen_for_order_book_snapshots(self.ev_loop, msg_queue)
+            )
+            self.async_run_with_timeout(self.listening_task)
+
+        self.assertEqual(0, msg_queue.qsize())
+
+    @aioresponses()
+    def test_listen_for_order_book_snapshots_logs_exception_error_with_response(self, mock_api):
+        url = web_utils.public_rest_url(CONSTANTS.SNAPSHOT_REST_URL, domain=self.domain)
+        regex_url = re.compile(f"^{url}".replace(".", r"\.").replace("?", r"\?"))
+
+        mock_response = {
+            "m": 1,
+            "i": 2,
+        }
+        mock_api.get(regex_url, body=json.dumps(mock_response), callback=self.resume_test_callback)
+
+        msg_queue: asyncio.Queue = asyncio.Queue()
+
+        self.listening_task = self.ev_loop.create_task(
+            self.data_source.listen_for_order_book_snapshots(self.ev_loop, msg_queue)
+        )
+
+        self.async_run_with_timeout(self.resume_test_event.wait())
+
+        self.assertTrue(
+            self._is_logged("ERROR", "Unexpected error occurred fetching orderbook snapshots. Retrying in 5 seconds...")
+        )
+
+    @aioresponses()
+    def test_listen_for_order_book_snapshots_successful(self, mock_api):
+        url = web_utils.public_rest_url(CONSTANTS.SNAPSHOT_REST_URL, domain=self.domain)
+        regex_url = re.compile(f"^{url}".replace(".", r"\.").replace("?", r"\?"))
+
+        mock_response = {
+            "error": None,
+            "id": 0,
+            "result": {
+                "orderbook_p": {
+                    "asks": [[87705000, 1000000], [87710000, 200000]],
+                    "bids": [[87700000, 2000000], [87695000, 200000]],
+                },
+                "depth": 30,
+                "sequence": 455476965,
+                "timestamp": 1583555482434235628,
+                "symbol": self.ex_trading_pair,
+                "type": "snapshot",
+            },
+        }
+        mock_api.get(regex_url, body=json.dumps(mock_response))
+
+        msg_queue: asyncio.Queue = asyncio.Queue()
+        self.listening_task = self.ev_loop.create_task(
+            self.data_source.listen_for_order_book_snapshots(self.ev_loop, msg_queue)
+        )
+
+        result = self.async_run_with_timeout(msg_queue.get())
+
+        self.assertIsInstance(result, OrderBookMessage)
+        self.assertEqual(OrderBookMessageType.SNAPSHOT, result.type)
+        self.assertTrue(result.has_update_id)
+        self.assertEqual(result.update_id, 455476965)
+        self.assertEqual(self.trading_pair, result.content["trading_pair"])
+
+    def test_listen_for_funding_info_cancelled_error_raised(self):
```

Comment:
Missing `test_listen_for_funding_info_logs_exception` and `test_listen_for_funding_info_successful`

Reply:
done

Reply:
I think this one is also still missing. `test_listen_for_subscriptions_successful` validates that diffs and trades are processed correctly, but not the funding info